server/jetstream_cluster.go

```diff
@@ -4172,8 +4434,79 @@ func (s *Server) jsClusteredStreamUpdateRequest(ci *ClientInfo, acc *Account, su
 			}
 			rg.Peers = peers
 		} else {
-			// We are deleting nodes here.
-			rg.Peers = rg.Peers[:newCfg.Replicas]
+			// We are deleting nodes here. We want to do our best to preserve the current leader.
+			// We have support now from above that guarantees we are in our own Go routine, so can
+			// ask for stream info from the stream leader to make sure we keep the leader in the new list.
+			var curLeader string
+			if !s.allPeersOffline(rg) {
+				// Need to release js lock.
+				js.mu.Unlock()
+				s.mu.Lock()
+				inbox := s.newRespInbox()
+				results := make(chan *StreamInfo, 1)
+				// Store our handler.
+				s.sys.replies[inbox] = func(sub *subscription, _ *client, _ *Account, subject, _ string, msg []byte) {
+					var si StreamInfo
+					if err := json.Unmarshal(msg, &si); err != nil {
+						s.Warnf("Error unmarshaling clustered stream info response:%v", err)
+						return
+					}
+					select {
+					case results <- &si:
+					default:
+						s.Warnf("Failed placing remote stream info result on internal channel")
+					}
+				}
+				s.mu.Unlock()
+
+				isubj := fmt.Sprintf(clusterStreamInfoT, ci.serviceAccount(), cfg.Name)
+				s.sendInternalMsgLocked(isubj, inbox, nil, nil)
+
+				const timeout = 2 * time.Second
+				notActive := time.NewTimer(timeout)
+				defer notActive.Stop()
+
+				select {
+				case <-s.quitCh:
+					break
+				case <-notActive.C:
+					s.Warnf("Did not receive stream info results for '%s > %s'", acc, cfg.Name)
+				case si := <-results:
+					if si.Cluster != nil {
+						// The leader here is the server name, but need to convert to internal name.
+						curLeader = string(getHash(si.Cluster.Leader))
+					}
+				}
+				// Clean up here.
+				s.mu.Lock()
+				if s.sys != nil && s.sys.replies != nil {
```

Comment:
We use this code in other places, so likely a copy-paste, but then that means, should we check for s.sys and s.sys.replies not being nil before trying to register the handler above? The `sys.replies` (and maybe s.sys) is set to nil on shutdown. So if there is a panic it would be on shutdown, so less critical.

Reply:
It always exists these days except as you point out on shutdown. We could check for it, this eventually (internal messaging) needs to be reworked of course.